### ISSUE TITLE
Can't seem to forecast high values and getting negative values as well

### ISSUE THREAD

**Issue by oren0e** [facebook/prophet#797]
Hi,
I am struggling to make a forecast using prophet package in R.
My main goal is to forecast this series, lets say it is requests for bus rides in a vicinity of an airport.
My data is hourly and I have number of arrivals (flights) as an external regressor, as well as temperatures (weather).
I have tried to play with the order of the fourier series for the weekly and daily seasonalities, as well as with the prior scales and the same for changepoints. I have tried also transforming my data with log transformation.
I am judging my forecast using "rolling-window one-step-ahead" method. Each time I increase the 1-day date to forecast by a day and also increase the starting date of the training data by a day. I do this for 20 hours, so eventually I have a vector with 20 errors - the mean of which is my metric of evaluation. In my situation underpredicting is worse so I am using a LINEX type of error function (but I also check the symetric case so I'll be able to calculate the MAPE).

Although it seems that the data is fit well in the plot, I get a MAPE of 40% no matter what I do (can't get it lower than that). 2 things bother me:
**1.** When I look at the data it seems that prophet cant capture high values in my data, max value in data is 90 or so, but in forecast is only 40.
**2.** I get negative values in `yhat`.
Regarding (2) I have tried to use log transformation before running the series of models (with `log(1+x)` because I have zeroes in my data as well). While it did seem to get (almost) rid of the higher negative values, the error metrics were just terrible, the MAPE was huge (I did transform everything back from logs as needed).

What am I missing and how can I improve the forecast?
I am attaching here 2 plots from one of the iterations in the rolling-window procedure (that is why there is only 1 red dot i.e. datapoint forecast). The external regressor in the components plot is the flight arrivals, the weather regressor disappeared from the plot somehow.

Last thing - my data is hourly data starting from 21/12/2018 to 05/01/2019. Initially I did not have records for each hour, so I inserted zeroes in those places, this is reasonable decision in my opinion because I know that when there are no records there was no activity (=0).
![arrivals](https://user-images.githubusercontent.com/33298364/50741322-6c709f00-1204-11e9-8949-e9d31a1a87e3.jpeg)
![arrivals_components](https://user-images.githubusercontent.com/33298364/50741323-6c709f00-1204-11e9-805f-9ce3fca7992c.jpeg)

**Comment by bletham** [comment]
I'm not too surprised to see that the those really high values are missed. Only 3 days have those really high values, and so the usual daily seasonality is not going to capture it. You can see a similar effect in the documentation, like the example here: https://facebook.github.io/prophet/docs/quick_start.html#r-api . Basically these points are considered outliers.

If they are real, then you'll need to use an extra regressor to add a spike on the days that should have spike so the model can learn the spike magnitude. Based on this time series alone, there doesn't seem to be any way for the model to know when there should be a big spike and so you'd need to add an indicator of that to the model.

Negative values can also happen if the multiplicative effect adds up to something bigger than -100%, which seems to be the case here. The best that can be done about this for the moment is to just rectify yhat in the forecast dataframe, that is, replace negative values with 0. (The proper solution is to use a Poisson link function, which is on the todo list)

For the missing weather regressor - in this plot, all of the external regressors are combined into a single panel (extra_regressors_multiplicative). If you want to see an individual extra regressor, you can use `plot_forecast_component`.

As for improving forecast performance, just to be sure: when you are making your forecast, are you including the future extra regressors in the future dataframe? If you join future dates with history to get the extra regressors, then they may be set to 0 for future dates which would cause bad performance.

**Comment by oren0e** [comment]
You are exactly right about using the Poisson, when do you think a new release will be out that will include this feature?

About the forecast - my whole data are historical data, so I am basically testing the model at this point. I treat part of the data as historical and the other part (1 observation and rolling this window forward) as the forecast "out of sample". By doing so I am able to calculate any type of error actually, otherwise I would not have the real value of `y` and could not calculate the error. So yes, my all extra regressors have their values also in the future data frame because essentially they are all historical data being left out for the testing part.

I have used the solution proposed in one of the issues here to add an additive effect of the extra regressor of flight arrivals only in weekends (or other 2-3 days with higher volume). It did improve the result (lower error), but still, the MAPE is high.

By the way - regarding the spikes, I thought it would capture these with the help of the holidays data-frame because these spikes happen on dates that I know about and which I have inputed in to the holidays data-frame.

**Comment by bletham** [comment]
OK, not filling in the extra regressors is an easy thing to miss but it sounds like you're doing that right.

The reason that identifying those days as holidays might not be working for you is because the holiday effect is applied uniformly across the day (as you can see in the components plot). In this case it is a multiplicative effect, and we want it to increase things on the high end enough to reach those high peaks during that part of the day (around 00:00 to 04:00), however this effect is applied uniformly through the day and so it would also then increase things on the low part of the day (06:00 to 12:00, for instance) by the same amount.

In reality what we have is that the peak is much higher, but the low part is more or less the same. This type of behavior isn't super well supported by the software at the moment. Basically, the shape of the daily seasonality is effected by the holiday. What you would really want to do here is have a separate daily seasonality that you fit for holidays, which you can do using extra regressors but it's quite a pain. There is an issue for making this easier (#538) which is one of the top priorities for the next version, but that issue does include the manual approach I mentioned before. In the meantime, you could consider just constructing an extra regressor that has a little bump during the peak part of the day on the holiday in the same that you see for your daily seasonality, and 0 during the low part of the day. Then the model would fit a multiplier coefficient for the size of the effect and could fit those large effects without messing up the low end.

In this case the trend is pretty simple and the seasonality does seem to be reasonably well captured, what is the MAPE for your in-sample data? I'm guessing it would also be close to 40%. I think MAPE is not a great measure of loss in this particular case (in that it does not provide a good representation of "typical" performance) because you have many points that are close to 0. So predicting 2 when it should have been 1 will give 100% error. You might consider mean absolute error.

Poisson link function is in #337, and will be done after we move predictions into Stan. We attempted that in the summer but ran into an upstream issue that was recently fixed, so hopefully that will be coming soon too.

**Comment by oren0e** [comment]
Hi, first of all I would like to thank you for the thorough answers.

Now if I understand your suggestion regarding the seasonality interaction with holidays correctly - you mean that I should add an additional regressor for **each** holiday on my list with the appropriate daily seasonality interaction? If so, that would imply adding a lot of regressors (and will be quite cumbersome in the code), can it be done in prophet? (or maybe I did not understand you correctly in the first place).

Regarding the MAPE you are right, but as I wrote in the original post, my metric of judgment is primarily the LINEX loss function (`L(e) = exp(ae) -ae -1`) where in my case `a` (a shape parameter) was taken to be negative (specifically -0.1) because in that way, underpredictive errors have an exponential cost while overpredictive errors have linear cost. This is because in my case, if I am wrong I would rather be wrong up and not down. I used the MAPE just to get a sense of the meaning of the error because with the LINEX function it is hard to tell.
One strange thing that happens is that once I reach a decent LINEX value with my "rolling-window" method, I would then make a forecast for lets say 20 hours ahead, and in some of the times I get up to 50-60% of the errors being negative, i.e., underpredicting. Whereas I thought that once I reach a low enough value with the "rolling-window" procedure, I would then ensure to get most of the errors positive, i.e., overpredicting.

**Comment by oren0e** [comment]
In this case..?

The hyperparameters affect the fit which in turn is directly related to the error. You are right that it is indeed very hard to take down the LINEX error, and sometimes the decreases are very small. What priors does prophet use?

**Comment by bletham** [comment]
The situation here is really similar to a regularized regression model like Lasso or ridge regression. Basically what we are minimizing is something like
```
(y_hat(theta) - y_true) ** 2 + \lambda * ||theta||^2
```
where `g(theta)` is a prior, and `\lambda` is the regularization parameter we're tuning to minimize LINEX. We are still minimizing squared error, but some values of `\lambda` cause squared-error-minimization to produce better or worse LINEX errors. But we're still minimizing squared error.

It's really the same thing here, e.g. the changepoint prior scale puts an L1 regularization on the trend change magnitudes.

The actual priors (and likelihood) are here:
https://github.com/facebook/prophet/blob/master/R/inst/stan/prophet.stan#L103

`tau` there, for instance, is `changepoint.prior.scale`.

**Comment by oren0e** [comment]
@bletham I am reopening this issue because now I have more data, and it seems that the problem of the model missing out high values still persists regardless of holidays, which I thought was the reason for the miss when I opened this issue. Treating these high values as outliers also does not make sense because they are too many as can be seen in the plot.

Following https://github.com/facebook/prophet/issues/434 I added an extra regressor for seasonality in specific days (which is really just adding another column of the value of the series when its on day x and 0 otherwise and then adding that as an extra regressor). This did seem to improve the error a little bit, but I still get that these high-values are missed.

I am petty new to bayesian time series analysis so I dont know if stationarity is an issue here, and if so, how is it being addressed in prophet (if at all)? Because when it comes to time-series analysis the first thing I am always testing is the issue of stationarity.

![model_plot](https://user-images.githubusercontent.com/33298364/54202630-1f959700-44d9-11e9-8b7f-4fbecf18df39.jpeg)
![model_components](https://user-images.githubusercontent.com/33298364/54202633-24f2e180-44d9-11e9-9eec-af78527d47bf.jpeg)

**Comment by bletham** [comment]
Like mentioned above, this time series is low-volume count data and so really would be properly modeled with a Poisson or negative binomial link function instead of the Gaussian link function that was actually use. We're working on that right now and I'm hopeful we'll get it working for the next version.

In the meantime, one specific issue caused by this is that there is a lot of upward skew in the noise. Prophet, of course, assumes, normal noise which has 0 skew. If you do a log transform (e.g. log(1 + y)) of the data before fitting the model, that will reduce the skew substantially and I think you'll get better model fit. Then just apply the inverse transform to yhat.

**Comment by oren0e** [comment]
I have done that previously without great success, but I log-transfered only y without the extra regressors, did I do wrong?

**Comment by bletham** [comment]
Log transforming extra regressors would probably make sense as well, though I'm not sure how much extra benefit you'll get from it if log transforming y didn't do much.

**Comment by Karan-GM** [comment]
Hi,

I am facing same issue as yours. Prophet is unable to forecast the magnitude of high values required. Were you able to handle the problem? What approach did you consider and what worked?

**Comment by reidwil** [comment]
I'm also interested in @Karan-GM's follow up question. Can you point to a solution @bletham. 

Also, thanks a lot for making this and spending the time to answer and problem solve with those using your tool

**Comment by XingLe0917** [comment]
Yes. I faced the same issue as yours. These are many high values required unable to forecast. Please have a look at my result. And I try remove weekend dates, but the result also can't forecast the high valus. 
<img width="827" alt="WX20201124-110558@2x" src="https://user-images.githubusercontent.com/14801171/100041915-4d557780-2e45-11eb-9788-34d2057145b8.png">

**Comment by bletham** [comment]
@XingLe0917 it looks like there might be a lot of positive skew in the noise, in which case a log transform would be really helpful. Basically take the log of the data and then fit the model. Can you see how that works?

**Comment by philipandreadis** [comment]
@bletham Hello, I am experiencing a similar problem with my data and I have opened a new issue https://github.com/facebook/prophet/issues/1791#issue-792284890 for it. It is true that fitting the log transformed data is better but I am interested in predicted the original values and by transforming the logged predictions back to the original scale with exp() the problem persists.
